core/storage/statistics/gae_models.py

```diff
@@ -1134,6 +1138,115 @@ def delete_playthroughs_multi(cls, playthrough_ids):
         cls.delete_multi(instances)
 
 
+class LearnerAnswerDetailsModel(base_models.BaseModel):
+    """Model for storing the answer details, learner enters when they
+    are asked for it.
+
+    The id of instances of this class has the form
+        [ENTITY_TYPE].[ENTITY_ID].[GENERATED_STRING]
+    """
+    # ID of the entity.
+    entity_id = ndb.StringProperty(required=True, indexed=True)
+    # The type of entity, that whether it is state, question, or any other
+    # object if it is added in future.
+    entity_type = ndb.StringProperty(required=True, indexed=True)
+    # The answer submitted by the learner.
+    learner_answer = ndb.StringProperty(required=True, indexed=False)
+    # The response submitted by the learner.
+    learner_response = ndb.StringProperty(required=True, indexed=False)
+    # The time at which the response was created.
+    created_on = ndb.DateTimeProperty(required=True, indexed=False)
+    # Whether the response received has been resolved.
+    is_resolved = ndb.BooleanProperty(default=False, required=True)
+
+    @classmethod
+    def get_entity_id(cls, exploration_id, state_name):
+        """Gets entity_id for a batch model based on given exploration state.
+
+        Args:
+            exploration_id: str. ID of the exploration currently being played.
+            state_name: str. The name of the state.
+
+        Returns:
+            str. Returns entity_id for a new instance of this class.
+        """
+        return '%s.%s' % (exploration_id, state_name)
+
+    @classmethod
+    def generate_new_answer_details_id(cls, entity_type, entity_id):
+        """Generates a new answer detail ID which is unique.
+
+        Args:
+            entity_type: str. The type of the entity.
+            entity_id: str. The ID of the entity.
+
+        Returns:
+            str. An answer detail ID that is different from the IDs of all
+                the existing answer details within the given entity.
+
+        Raises:
+           Exception: There were too many collisions with existing answer
+               detail IDs when attempting to generate a new answer detail ID.
+        """
+        for _ in range(_MAX_RETRIES):
+            answer_details_id = (
+                entity_type + '.' + entity_id + '.' +
+                utils.base64_from_int(utils.get_current_time_in_millisecs()) +
+                utils.base64_from_int(utils.get_random_int(_RAND_RANGE)))
+            if not cls.get_by_id(answer_details_id):
+                return answer_details_id
+        raise Exception(
+            'New answer detail id generator is producing too many collisions.')
+
+    @classmethod
+    def create(
+            cls, exp_id, state_name, entity_type, learner_answer,
```

Comment:
Ditto here, exp_id and state_name parameters are exclusive to an exploration. Just pass entity_id here and the calling function should take care of passing in the correct entity_id.

Reply:
Done